src/soc/rockchip/rk3399.c

```diff
@@ -0,0 +1,510 @@
+		return -1;
+	}
+
+	return soc_wait_for_interrupt(rk3399, pin->fd, ms);
+}
+
+
+
+static int rk3399GC(void) {
+	struct layout_t *pin = NULL;
+	char path[PATH_MAX];
+	int i = 0;
+
+	if(rk3399->map != NULL) {
+		for(i=0;i<rk3399->map_size;i++) {
+			pin = &rk3399->layout[setPin(rk3399->map[i],rk3399->map)];
+			if(pin->mode == PINMODE_OUTPUT) {
+				pinMode(i, PINMODE_INPUT);
+			} else if(pin->mode == PINMODE_INTERRUPT) {
+				sprintf(path, "/sys/class/gpio/gpio%d", rk3399->irq[i]);
+				if((soc_sysfs_check_gpio(rk3399, path)) == 0) {
+					sprintf(path, "/sys/class/gpio/unexport");
+					soc_sysfs_gpio_unexport(rk3399, path, rk3399->irq[i]);
+				}
+			}
+			if(pin->fd > 0) {
+				close(pin->fd);
+				pin->fd = 0;
+			}
+		}
+	}
+	if(rk3399->gpio[0] != NULL) {
+		munmap(rk3399->gpio[0], rk3399->page_size);
+	}
+	return 0;
+}
+
+
+static int rk3399SelectableFd(int i) {
+	struct layout_t *pin = NULL;
+
+	if(rk3399->irq == NULL) {
+		wiringXLog(LOG_ERR, "The %s %s has not yet been mapped", rk3399->brand, rk3399->chip);
+		return -1;
+	}
+	if(rk3399->fd <= 0 || rk3399->gpio == NULL) {
+		wiringXLog(LOG_ERR, "The %s %s has not yet been setup by wiringX", rk3399->brand,rk3399->chip);
+		return -1;
+	}
+
+	pin = &rk3399->layout[rk3399->irq[i]];
+	return pin->fd;
+}
+
+static int rk3399I2CSetup(int devId){
+    int fd;
+    const char *device = NULL;
+
+    device = "/dev/i2c-7";
+    if((fd = open(device,O_RDWR)) < 0){
+        wiringXLog(LOG_ERR,"cannot open %s %s's %s",rk3399->brand,rk3399->chip,device);
+        return -1;
+    }
+    if(ioctl(fd,I2C_SLAVE,devId) < 0){
+        wiringXLog(LOG_ERR,"Unable to set %s %s's %s slave mode",rk3399->brand,rk3399->chip,device);
+        return -1;
+    }
+    return fd;
+}
+
+
+static int rk3399I2CRead(int fd){
+    return i2c_smbus_read_byte(fd);
+}
+
+static int rk3399I2CReadReg8(int fd, int cmd) {
+	return i2c_smbus_read_byte_data(fd, cmd);
+}
+
+static int rk3399I2CReadReg16(int fd, int cmd) {
+	return i2c_smbus_read_word_data(fd, cmd);
+}
+
+static int rk3399I2CWrite(int fd, int value) {
+	return i2c_smbus_write_byte(fd, value);
+}
+
+static int rk3399I2CWriteReg8(int fd, int cmd, int value) {
+	return i2c_smbus_write_byte_data(fd, cmd, value);
+}
+
+static int rk3399I2CWriteReg16(int fd, int cmd, int value) {
+	return i2c_smbus_write_word_data(fd, cmd, value);
+}
+
+void rk3399Init(void) {	
+    soc_register(&rk3399, "Rockchip", "Rk3399");
+
+	rk3399->layout = layout;
+
+	rk3399->support.isr_modes = ISR_MODE_RISING | ISR_MODE_FALLING | ISR_MODE_BOTH | ISR_MODE_NONE;
+
+	rk3399->page_size = (4*1024);
+	//rk3399->base_addr[0] = GPIO0_BASE;
+    rk3399->base_addr[1] = GPIO1_BASE;
+    rk3399->base_addr[2] = GPIO2_BASE;
+    rk3399->base_addr[3] = GPIO3_BASE;
+    rk3399->base_addr[4] = GPIO4_BASE;
+	rk3399->base_offs[0] = 0x00000000;
+
+	rk3399->gc = &rk3399GC;
+	rk3399->selectableFd = &rk3399SelectableFd;
+
+	rk3399->pinMode = &rk3399PinMode;
+    rk3399->setup = &rk3399Setup;
```

Comment:
Indentation

Reply:
thanks for your suggestions, I'll fix those right away